![](https://www.dii.uchile.cl/wp-content/uploads/2021/06/Magi%CC%81ster-en-Ciencia-de-Datos.png)

**MDS7202: Laboratorio de Programación Científica para Ciencia de Datos**

### 👨‍🏫👩‍🏫 Cuerpo Docente:

- Profesores: Diego Cortez, Gabriel Iturra
- Auxiliares: Melanie Peña, Valentina Rojas
- Ayudantes: Nicolás Cabello, Cristopher Urbina

### 👨‍💻👩‍💻 Estudiantes:
- Estudiante n°1: Daniel Vargas
- Estudiante n°2: Rafael Tapia

_Por favor, lean detalladamente las instrucciones de la tarea antes de empezar a escribir._

---

### **Instrucciones importantes**

1. **Formato del informe**:  
   - El informe debe estar integrado dentro de un **Jupyter Notebook**. No es necesario subirlo a una plataforma externa, pero debe cumplir con los siguientes requisitos:  
     - Estructura clara y ordenada.  
     - Código acompañado de explicaciones detalladas.  
     - Resultados presentados de forma visual y analítica.  

2. **Descuento por informes deficientes**:  
   - Cualquier sección del informe que no tenga una explicación adecuada o no respete el formato será penalizada con un descuento en la nota. Esto incluye código sin comentarios o análisis que no sean coherentes con los resultados presentados.
   - **Comentarios sin formatear de ChatGPT o herramientas similares serán penalizados (e.g: "Inserta tu modelo acá", etc.)**

<center>
<img src="https://media1.giphy.com/media/ZkIUY6LZJMwrSMQMWu/giphy.gif?cid=6c09b952msc755426bpsadn4sysofcpqczczi7x1pghd41sw&ep=v1_internal_gif_by_id&rid=giphy.gif&ct=v" width="400" height="300">

# 📬 Entrega Parcial 3 (30% del Proyecto)

### 📪 Fecha de Entrega: 

Esta entrega cuenta con 2 partes:

- **Predicciones**: 1 de Diciembre, 2 de Diciembre, 3 de Diciembre, 4 de Diciembre
- **Informe:** 8 de Diciembre.

## 📖 Enunciado

Después de haber superado con éxito la implementación de sus flujos de trabajo con `Airflow`, el equipo **Deep Drinkers 🤖** ha demostrado que no solo sabe construir modelos, sino también integrarlos en pipelines reales listos para producción.

El equipo técnico de **SodAI Drinks 🥤** ha quedado tan satisfecho con la automatización de sus procesos que ahora les ha encomendado una nueva, y final, misión:  
- 📊 **Evaluar la calidad real de sus modelos en un entorno más desafiante** 
- 🧠 **Comunicar los hallazgos de forma clara y profesional frente a una comisión**.

<center>
<img src="https://i.giphy.com/btbUGSHh3f6eBjbDfh.webp" width="350">


## 📌 Predicciones
<center>
<img src="https://i.giphy.com/NTMxntb8rMzmbd1x97.webp" width="500" height="300">
</center>

Su *pipeline* automatizado debe demostrar que puede manejar datos reales en producción. **SodAI Drinks** 🥤 liberará *batches* de datos en fechas diferentes para evaluar si su sistema se adapta correctamente, los cuales simularán la llegada de información nueva (cada batch es una semana).

Lo que debe funcionar automáticamente:
- **Ejecución Automatizada**: Su DAG de `Airflow` debe ser capaz de procesar cada *batch* de datos de manera automática cuando esté disponible.
- **Monitoreo Continuo**: El *pipeline* debe evaluar la calidad de los datos entrantes y detectar posibles desviaciones respecto al conjunto original.
- **Decisión de Reentrenamiento**: Basándose en los análisis de *data drift* y métricas de rendimiento, el sistema debe determinar automáticamente si es necesario reentrenar el modelo.
- **Tracking Completo**: Todas las ejecuciones, métricas y decisiones deben quedar registradas en `MLFlow` para posterior análisis.
- **Generación de Predicciones**: Para cada *batch* procesado, el sistema debe generar las predicciones correspondientes y almacenarlas de manera organizada.

**IMPORTANTE: Es imperativo que guarden los resultados de su pipeline (métricas, tracking, predicciones, etc) para su análisis en la siguiente sección.**

### 📌Entregable: Competencia en CodaLab

<center>
<img src="https://i.giphy.com/NomCzPIGoXs3EIq77v.webp" width="300" height="300">
</center>

Como parte de esta entrega final, el equipo debe utilizar su pipeline entrenado para **generar predicciones en los nuevos conjuntos de datos mencionados anteriormente**.

Estas predicciones deben:

- Ser generadas directamente desde el *pipeline* previamente desarrollado.  
- Guardarse en archivos `.csv` siguiendo el formato requerido.  
- Subirse a la plataforma **CodaLab**, donde se realizará la evaluación final.

En CodaLab podrán:

- Ver el rendimiento de su modelo frente a los de otros equipos.  
- Obtener una puntuación basada en la métrica definida del proyecto.  

Cada set de predicciones debe ser publicado en CodaLab en las siguientes fechas:
- **Predicciones 1 📅** : 1 de Diciembre (Fecha de predicción: 01/01/25 - 05/01/25)
    - Note que para este set de predicciones, sólo tendrán acceso a los datos publicados en la Entrega 1.
- **Predicciones 2 📅** : 2 de Diciembre (Fecha de predicción: 06/01/25 - 12/01/25)
- **Predicciones 3 📅** : 3 de Diciembre (Fecha de predicción: 13/01/25 - 19/01/25)
- **Predicciones 4 📅** : 4 de Diciembre (Fecha de predicción: 20/01/25 - 26/01/25)

Para simular la llegada de nuevos datos, se publicarán los siguientes conjuntos de datos:
- **Batch 1 📅** : 2 de Diciembre (Con datos realizados del período 01/01/25 - 05/01/25)
- **Batch 2 📅** : 3 de Diciembre (Con datos realizados del período 06/01/25 - 12/01/25)
- **Batch 3 📅** : 4 de Diciembre (Con datos realizados del período 13/01/25 - 19/01/25)
- **Batch 4 📅** : 5 de Diciembre (Con datos realizados del período 20/01/25 - 26/01/25)
    - Note como este conjunto de datos se publica **después** de la competencia, por lo que sólo les servirá para la última evaluación requerida en el informe.

**IMPORTANTE:** Subir los resultados a tiempo en las tres fechas es **<u>OBLIGATORIO</u>** para la evaluación del desempeño final del equipo. **Por cada fecha en la que no se suban predicciones, se aplicará un descuento de 0.75 puntos (75 décimas) sobre la nota de la Entrega 3.**

### 🎁 Bonus [0.5 puntos]

Como incentivo adicional, se premiará a los **3 equipos con mejor performance en la métrica F1** con **0.5 puntos de puntaje adicional** sobre la nota de la Entrega 3.

## 📌 Informe: Performance del modelo [6.0 puntos]

En esta sección se espera que el equipo analice de manera retrospectiva el performance de su modelo y saque conclusiones en función de sus resultados. 

Para esto, tendrán que incurrir en 3 tipos de análisis:

- **Individual:** Deben evaluar el performance de su modelo tomando cada semana de forma aislada.

- **Comparativo:** Deben evaluar el performance de su modelo *a través* de las semanas, comparando el desempeño entre las semanas e identificando posibles tendencias.
    - Para esta parte se espera que generen gráficos de tendencia y tablas comparativas para apoyar su análisis.

- **Conclusiones y Aprendizajes:** Deben escribir sus principales conclusiones y aprendizajes de este proyecto. 

A lo largo de esta sección, se espera que respondan preguntas como:

1. ¿Cómo variaron sus métricas a lo largo de los distintos conjuntos de datos? 
2. ¿En qué momento el modelo tuvo su peor desempeño y por qué?
3. ¿Detectaron algún cambio significativo (drift) en la distribución de los datos? ¿Cómo lo identificaron?  
4. ¿Tuvieron que reentrenar su modelo con los nuevos datos? ¿Porqué? ¿Ayudó esto al performance de su modelo?
4. ¿Qué decisión técnica (modelo, métrica, imputación, etc.) tuvo más impacto en los resultados?  
5. ¿Qué hiperparámetro fue el más importante para su modelo? 
5. ¿Qué variable fue más influyente en las predicciones? ¿Cómo lo interpretan? ¿Cómo cambió su importancia con respecto a las otras variables a lo largo de los batch de datos?
6. ¿Qué aprendieron sobre el negocio a partir de los resultados del modelo?
7. ¿Qué limitaciones detectaron en su modelo o en los datos?  

**IMPORTANTE: Se espera que en sus respuestas hagan referencia a los artefactos (métricas, hiperparámetros, gráficos de interpretabilidad, etc) que su pipeline genera.**

<center>
<img src="https://media1.tenor.com/m/jyQ3SPT1htEAAAAd/i-love-this-performance-even-more-simon-cowell.gif" width="300" height="300">
</center>

### 📌 Análisis individual [3.0 puntos]

In [44]:
import pandas as pd

clientes_df = pd.read_parquet('data/clientes.parquet')
productos_df = pd.read_parquet('data/productos.parquet')

# Calcula el tamaño del universo total (Clientes x Productos) para el True Negative (TN)
universe_size = len(clientes_df) * len(productos_df)

def get_batch_performance(batch_id):
    # Carga de datos reales (Ground Truth) y creación de sets de pares
    actual_df = pd.read_parquet(f"data/batch_{batch_id}/batch_t{batch_id}.parquet")
    actual_pairs = set(map(tuple, actual_df[['customer_id', 'product_id']].values))
    
    # Carga de predicciones y creación de sets de pares
    preds_df = pd.read_csv(
        f"predictions/batch_{batch_id}/predictions_next_week.csv",
        header=None,
        names=['customer_id', 'product_id']
    )
    pred_pairs = set(map(tuple, preds_df[['customer_id', 'product_id']].values))
    
    # Cálculo de la matriz de confusión usando operaciones de conjuntos
    tp = len(pred_pairs & actual_pairs) # True Positives (Aciertos)
    fp = len(pred_pairs - actual_pairs) # False Positives (Ruido)
    fn = len(actual_pairs - pred_pairs) # False Negatives (Omisiones)
    # True Negatives se calcula por diferencia respecto al universo total
    tn = universe_size - tp - fp - fn
    
    predictions_size = len(pred_pairs)
    total_compras_reales = len(actual_pairs)

    # Cálculo de métricas globales
    precision = tp / predictions_size if predictions_size > 0 else 0
    recall = tp / total_compras_reales if total_compras_reales > 0 else 0
    f1 = 2 * (precision * recall) / (precision + recall) if (precision + recall) > 0 else 0
    accuracy = (tp + tn) / universe_size

    # Marcando si cada predicción fue un acierto o no para el análisis desagregado
    preds_df['is_hit'] = preds_df.apply(lambda x: (x['customer_id'], x['product_id']) in actual_pairs, axis=1)
    
    # Agrupación por cliente para obtener estadísticas individuales
    user_stats = preds_df.groupby('customer_id').agg(
        total_recs=('product_id', 'count'),
        hits=('is_hit', 'sum')
    )
    user_stats['user_precision'] = user_stats['hits'] / user_stats['total_recs']
    
    # Cálculo de métricas de distribución por usuario para el informe
    total_users_with_recs = len(user_stats)
    pct_precision_zero = (len(user_stats[user_stats['user_precision'] == 0]) / total_users_with_recs) * 100
    pct_precision_gt_10 = (len(user_stats[user_stats['user_precision'] > 0.10]) / total_users_with_recs) * 100
    
    metrics = {
        "global": {
            "accuracy": accuracy,
            "precision": precision,
            "recall": recall,
            "f1": f1,
            "predictions_count": predictions_size,
            "tp": tp, 
            "fp": fp, 
            "fn": fn
        },
        "user_stats": {
            "pct_users_precision_zero": pct_precision_zero,
            "pct_users_precision_gt_10": pct_precision_gt_10,
            "avg_recs_per_user": user_stats['total_recs'].mean(),
            "max_recs_per_user": user_stats['total_recs'].max()
        }
    }

    return metrics

Para complementar el análisis de las métricas globales, se incorporaron métricas de desempeño a nivel de cliente (user_stats). 

Este análisis se utiliza para evaluar la equidad y la robustez del modelo a través de diferentes segmentos de usuarios, ya que los promedios globales pueden ocultar fallos sistémicos para grupos específicos.

Las métricas clave agregadas y su significado son:

pct_users_precision_zero (Porcentaje de Usuarios con Precisión Cero): Indica la proporción de clientes que recibieron sugerencias pero donde ninguna de ellas resultó en una compra real (acierto). Un valor alto en esta métrica revela un grupo para el cual el modelo predice aleatoriamente o carece de información útil, lo que requiere estrategias de filtrado o modelos segmentados.

pct_users_precision_gt_10 (Porcentaje de Usuarios con Precisión Mayor al 10%): Señala el tamaño del segmento de clientes para el cual el modelo es consistente y efectivo. Este grupo representa el target ideal donde las recomendaciones funcionan bien (definiendo bien como 10% de precisión).

avg_recs_per_user y max_recs_per_user: Cuantifican la cantidad de productos máxima y promedio por cliente. Estos valores son para definir políticas de Top-N y asegurar que la cantidad de sugerencias emitida sea viable y no sature al usuario final (definidas desde el error ocurrido por el negative sampling).

#### 📌 Batch 1 (01/01/25 - 05/01/25) [0.75 puntos]

In [45]:
import pandas as pd

clientes_df = pd.read_parquet('data/clientes.parquet')
productos_df = pd.read_parquet('data/productos.parquet')
universe_size = len(clientes_df) * len(productos_df)
transacciones_df = pd.read_parquet('./data/transacciones.parquet')


print(f"Total datos: {len(transacciones_df):,} transacciones")

metrics = get_batch_performance(1)
print(f"\nMétricas Batch 1:")
for section, data in metrics.items():
    print(f"\n {section.upper()} ")
    for k, v in data.items():
          print(f"  {k}: {v:.4f}")

Total datos: 254,936 transacciones

Métricas Batch 1:

 GLOBAL 
  accuracy: 0.9957
  precision: 0.2747
  recall: 0.6195
  f1: 0.3806
  predictions_count: 7361.0000
  tp: 2022.0000
  fp: 5339.0000
  fn: 1242.0000

 USER_STATS 
  pct_users_precision_zero: 52.6929
  pct_users_precision_gt_10: 47.0888
  avg_recs_per_user: 5.3574
  max_recs_per_user: 24.0000


Tras la optimización de hiperparámetros, el modelo generó un total de 7.361 sugerencias. Esta cifra significa una reducción de 23 veces respecto a las aprox. 170.000 predicciones de la iteración anterior. Esta reducción se debe a la corrección del parámetro scale_pos_weight mediante Optuna, lo que eliminó el inflado artificial de falsos positivos pues se disminuyo mucho el valor del parámetro scale_pos_weight pues se calculaba en función de pos/negs que para este caso es demasiado grande.

Al comparar el nuevo conjunto con las compras efectivas, se registraron 2022 verdaderos positivos (TP), 5339 falsos positivos (FP) y 1242 falsos negativos (FN). Con estos valores, se alcanzó una precisión de 0,2747 y un recall de 0,6195.

A diferencia del modelo anterior, que emitía casi 57 recomendaciones por cada compra real (saturando al cliente), la versión actual optimizada con Optuna emite aproximadamente 3,6 recomendaciones por cada interacción verdadera ($1 / 0.2747$). Esto eleva el F1-score de 0,12 a 0,3806, indicando un balance mucho mejor entre cobertura y precisión.

El análisis desagregado (de user stats) muestra que el 52.7% de los usuarios con recomendaciones tiene una precisión igual a cero, mientras que un 47.1% supera el 10% de aciertos. Aunque la mejora global es evidente, esta variabilidad sugiere que todavía existe un subgrupo de usuarios para los cuales el modelo no logra capturar patrones de compra claros.

La corrección en el entrenamiento, específicamente al delegar el ajuste del scale_pos_weight a Optuna en lugar de usar la proporción directa de clases, eliminó el ruido masivo de falsos positivos.

#### 📌 Batch 2 (06/01/25 - 12/01/25) [0.75 puntos]

In [46]:
import pandas as pd

clientes_df = pd.read_parquet('data/clientes.parquet')
productos_df = pd.read_parquet('data/productos.parquet')
universe_size = len(clientes_df) * len(productos_df)
transacciones_df = pd.read_parquet('./data/transacciones.parquet')


print(f"Total datos: {len(transacciones_df):,} transacciones")

metrics = get_batch_performance(2)
print(f"\nMétricas Batch 2:")
for section, data in metrics.items():
    print(f"\n {section.upper()} ")
    for k, v in data.items():
          print(f"  {k}: {v:.4f}")

Total datos: 254,936 transacciones

Métricas Batch 2:

 GLOBAL 
  accuracy: 0.9954
  precision: 0.3807
  recall: 0.5859
  f1: 0.4616
  predictions_count: 7853.0000
  tp: 2990.0000
  fp: 4863.0000
  fn: 2113.0000

 USER_STATS 
  pct_users_precision_zero: 32.8652
  pct_users_precision_gt_10: 66.9944
  avg_recs_per_user: 5.5147
  max_recs_per_user: 21.0000


El modelo fue reentrenado incorporando los datos del Batch 1, generando un total de 7.853 sugerencias. 

Este nuevo ciclo de entrenamiento resultó en un incremento notable de la Precisión a 0,3807 (frente a 0,2747 en el Batch 1) y un $\text{F1-score}$ de 0,4616. La mejora en precisión significa que de cada 100 recomendaciones emitidas, 38 fueron correctas.A nivel de la matriz de confusión, se obtuvieron 2.990 verdaderos positivos (TP) y 4.863 falsos positivos (FP). 

El Recall (Cobertura) se situó en 0,5859, indicando que se recuperó el 58,6% del total de las compras reales. 

Operacionalmente, el sistema emite ahora aproximadamente 2,6 recomendaciones por cada interacción verdadera ($\frac{1}{0,3807}$), lo que representa una mayor eficiencia en la comunicación con el cliente.

El análisis desagregado por usuario muestra que hay una distribución mejor del rendimiento del modelo porque La proporción de usuarios con Precisión Cero (pct_users_precision_zero) se redujo drásticamente al 32,87% (frente al 52,69% del Batch 1).

Consecuentemente, el porcentaje de clientes que superan el 10% de aciertos (pct_users_precision_gt_10) se incrementó significativamente, alcanzando el 66,99%.

#### 📌 Batch 03 (13/01/25 - 19/01/25) [0.75 puntos]

In [47]:
import pandas as pd

clientes_df = pd.read_parquet('data/clientes.parquet')
productos_df = pd.read_parquet('data/productos.parquet')
universe_size = len(clientes_df) * len(productos_df)
transacciones_df = pd.read_parquet('./data/transacciones.parquet')


print(f"Total datos: {len(transacciones_df):,} transacciones")

metrics = get_batch_performance(3)
print(f"\nMétricas Batch 3:")
for section, data in metrics.items():
    print(f"\n {section.upper()} ")
    for k, v in data.items():
          print(f"  {k}: {v:.4f}")

Total datos: 254,936 transacciones

Métricas Batch 3:

 GLOBAL 
  accuracy: 0.9922
  precision: 0.2547
  recall: 0.7367
  f1: 0.3785
  predictions_count: 14239.0000
  tp: 3626.0000
  fp: 10613.0000
  fn: 1296.0000

 USER_STATS 
  pct_users_precision_zero: 34.2992
  pct_users_precision_gt_10: 63.4771
  avg_recs_per_user: 9.5950
  max_recs_per_user: 32.0000


El modelo fue nuevamente reentrenado incorporando los datos de los batches anteriores. Esta iteración muestra un cambio notable en la estrategia de predicción, priorizando la cobertura sobre la precisión, lo que resulta en un aumento significativo en el volumen de recomendaciones.

La última iteración produjo un aumento significativo en el volumen, generando $14.239$ predicciones. Este cambio priorizó la Cobertura

El $\text{Recall}$ alcanzó su punto más alto en $0,7367$ (capturando $73,7\%$ de las compras reales).

Esta cobertura se logró a costa de la Precisión, que cayó a $0,2547$, el valor más bajo de la serie.

La eficiencia se deterioró, volviendo a requerir $3,9$ recomendaciones por acierto. Además, la carga operativa subió, con un promedio de $9,60$ recomendaciones por usuario y un máximo de $32,00$, incrementando el riesgo de fatiga del cliente.

#### 📌 Batch 04 (20/01/25 - 26/01/25) [0.75 puntos]

In [50]:
import pandas as pd

def analyze_prediction_distribution(batch_id):
    #Leer csv de predicciones
    preds_df = pd.read_csv(
        f"predictions/batch_{batch_id}/predictions_next_week.csv",
        header=None,
        names=['customer_id', 'product_id']
    )
    
    # Análisis de distribución de predicciones
    total_preds = len(preds_df)
    unique_users = preds_df['customer_id'].nunique()
    unique_products = preds_df['product_id'].nunique()
    
    user_stats = preds_df.groupby('customer_id').size()
    
    # Cálculo de estadísticas
    stats = {
        "volume": {
            "total_predictions": total_preds,
            "unique_users_targeted": unique_users,
            "unique_products_recommended": unique_products
        },
        "distribution": {
            "avg_recs_per_user": user_stats.mean(),
            "median_recs_per_user": user_stats.median(),
            "min_recs_per_user": user_stats.min(),
            "max_recs_per_user": user_stats.max(),
            "std_recs_per_user": user_stats.std()
        }
    }
    return stats

stats = analyze_prediction_distribution(4)

for section, data in stats.items():
    print(f"\n--- {section.upper()} ---")
    for k, v in data.items():
        if isinstance(v, float):
             print(f"  {k:<30}: {v:.4f}")
        else:
             print(f"  {k:<30}: {v:,}")


--- VOLUME ---
  total_predictions             : 13,742
  unique_users_targeted         : 1,474
  unique_products_recommended   : 68

--- DISTRIBUTION ---
  avg_recs_per_user             : 9.3229
  median_recs_per_user          : 8.0000
  min_recs_per_user             : 1
  max_recs_per_user             : 32
  std_recs_per_user             : 5.7924


Para el cuarto ciclo de predicción, al no disponer aún de las transacciones efectivas para validación (periodo de venta en curso), se realizó un análisis de distribución de las sugerencias emitidas para asegurar la estabilidad y prevenir la saturación del cliente.

El modelo generó un total de 13.742 sugerencias dirigidas a 1.474 clientes únicos. El catálogo activo en las recomendaciones abarcó 68 productos distintos.

El análisis de carga muestra un comportamiento expansivo, muy similar al observado en el Batch 3. El promedio de recomendaciones por usuario se situó en 9,32, con una mediana de 8 sugerencias. El pico de carga se mantiene alto, enviando un máximo de 32 sugerencias a un solo cliente.

La distribución de las predicciones del Batch 4 replica casi con exactitud el patrón del Batch 3 (que tuvo un promedio de 9,60). Esto indica que el modelo mantiene una estrategia de alta cobertura, priorizando el volumen sobre la selectividad observada en el Batch 2. Dado que este volumen conlleva un mayor riesgo de fatiga del cliente (saturación por exceso de oferta), se recomienda aplicar filtros estrictos (Top-N) antes del envío final si se desea priorizar la experiencia de usuario sobre la cobertura bruta.

Para esta última fase aunque no se cuenta con el ground truth, en el test de codabench se llego a un score de 0.39 mejorando ligeramente con respecto a la última.

### 📌 Análisis comparativo [2.0 puntos]

In [52]:
import pandas as pd
import os

BATCH_IDS = [1, 2, 3, 4]

def get_comparative_stats(batch_ids):
    
    report_data = []

    # Cargar Baseline (Predicciones del Batch 1)
    base_path = "predictions/batch_1/predictions_next_week.csv"
    base_df = pd.read_csv(base_path, header=None, names=['customer_id', 'product_id'])
    base_pairs = set(map(tuple, base_df[['customer_id', 'product_id']].values))
    base_count = len(base_pairs)

    for b_id in batch_ids:
        
        # Cargar Predicciones del Batch Actual
        pred_path = f"predictions/batch_{b_id}/predictions_next_week.csv"
            
        preds_df = pd.read_csv(pred_path, header=None, names=['customer_id', 'product_id'])
        pred_pairs = set(map(tuple, preds_df[['customer_id', 'product_id']].values))
        predictions_count = len(pred_pairs)
        
        # Cargar Ground Truth
        true_path = f"data/batch_{b_id}/batch_t{b_id}.parquet"
        has_truth = os.path.exists(true_path)
        
        metrics = {}
        
        # INFO GENERAL
        metrics['Batch'] = f"Batch {b_id}"
        metrics['Predicciones'] = predictions_count
        
        if has_truth:
            actual_df = pd.read_parquet(true_path)
            actual_pairs = set(map(tuple, actual_df[['customer_id', 'product_id']].values))
            total_real = len(actual_pairs)
            
            # RENDIMIENTO MODELO ACTUAL
            tp = len(pred_pairs & actual_pairs)
            precision = tp / predictions_count if predictions_count > 0 else 0
            recall = tp / total_real if total_real > 0 else 0
            f1 = 2 * (precision * recall) / (precision + recall) if (precision + recall) > 0 else 0
            
            # RENDIMIENTO BASELINE (PREDICCIONES BATCH 1)
            tp_base = len(base_pairs & actual_pairs)
            prec_base = tp_base / base_count if base_count > 0 else 0
            rec_base = tp_base / total_real if total_real > 0 else 0
            f1_base = 2 * (prec_base * rec_base) / (prec_base + rec_base) if (prec_base + rec_base) > 0 else 0

            # User stats
            user_counts = preds_df.groupby('customer_id').size()
            preds_df['is_hit'] = preds_df.apply(lambda x: (x['customer_id'], x['product_id']) in actual_pairs, axis=1)
            user_hits = preds_df.groupby('customer_id')['is_hit'].sum()
            user_precision = user_hits / user_counts
            pct_zero = (len(user_precision[user_precision == 0]) / len(user_precision)) * 100

            metrics['F1 Actual'] = f1
            metrics['F1 Baseline (B1)'] = f1_base
            metrics['Delta F1'] = f1 - f1_base
            metrics['Precision Actual'] = precision
            metrics['Precision Base'] = prec_base
            metrics['% Usr Prec 0'] = pct_zero
            
        else:
            # Sin datos reales (Batch 4)
            metrics['F1 Actual'] = None
            metrics['F1 Baseline (B1)'] = None
            metrics['Delta F1'] = None
            metrics['Precision Actual'] = None
            metrics['Precision Base'] = None
            metrics['% Usr Prec 0'] = None

        report_data.append(metrics)

    return pd.DataFrame(report_data)

# Ejecutar análisis
df_comparativo = get_comparative_stats(BATCH_IDS)

# Visualización
cols_show = [
    'Batch', 
    'F1 Actual', 'F1 Baseline (B1)', 'Delta F1',
    'Precision Actual', 'Precision Base', 
    '% Usr Prec 0'
]

print(df_comparativo[cols_show].to_string(index=False, float_format="%.4f"))

  Batch  F1 Actual  F1 Baseline (B1)  Delta F1  Precision Actual  Precision Base  % Usr Prec 0
Batch 1     0.3806            0.3806    0.0000            0.2747          0.2747       52.6929
Batch 2     0.4616            0.4677   -0.0062            0.3807          0.3960       32.8652
Batch 3     0.3785            0.4645   -0.0861            0.2547          0.3876       34.2992
Batch 4        NaN               NaN       NaN               NaN             NaN           NaN


El análisis comparativo revela un comportamiento contraintuitivo, las predicciones generadas por el modelo original (usadas para el Batch 1) superaron consistentemente en rendimiento a las versiones reentrenadas. Se observa un Delta F1 negativo progresivo, comenzando con una leve disminución de -0,0062 en el Batch 2 y acentuándose drásticamente en el Batch 3 con una caída de -0,0861 (donde el modelo base mantuvo un F1 de 0,46 frente al 0,37 del actual). Este deterioro en las métricas sugiere que la estrategia de reentrenamiento incremental podría estar introduciendo ruido o sobreajuste en lugar de aprendizaje útil, o que se implemento de manera incorrecta el reentrenamiento a nivel de código.

### 📌 Conclusiones y Aprendizajes [1.0 puntos]

La implementación de la optimización de hiperparámetros para calibrar el scale_pos_weight fue el mayor avance a nivel de rendimiento en f1, logrando estabilizar el desbalance de clases.

Sin embargo, ocurrió un problema al momento de agregar cada batch, la degradación por reentrenamiento continuo. Los datos demuestran que el modelo original (Batch 1) generalizó mejor en periodos futuros que sus versiones actualizadas, las cuales mostraron una tendencia a priorizar la cobertura desmedida (sacrificando precisión) a medida que ingerían más datos de nuevos periodos.

Como lección principal, se concluye que el volumen de datos no garantiza calidad predictiva. Para las siguientes iteraciones, es imperativo auditar la estrategia de ingestión de datos y priorizar el uso del modelo base estabilizado, además es importante la revisión del código de forma exhaustiva para comprobar que el motivo de esto no sea la forma en que estos nuevos datos se usaron.

## 📁 Estructura de entrega

Se espera que su entrega cuente con la siguiente estructura de carpetas:

```bash
entrega_3/
│
├── informe.ipynb # Jupyter notebook con sus resultados
│
├── predictions/ # Carpeta con las predicciones para cada batch de datos
│
└── data/ # Carpeta con los datos realizados (lo que realmente ocurrió)
```

Mucho éxito!

<center>
<img src="https://gifdb.com/images/high/may-the-force-be-with-you-anakin-skywalker-n2g5o0pm4h6iylsx.gif" width="400" height="300">
